content/200-concepts/100-components/02-prisma-client/052-extend-client.mdx

```diff
@@ -0,0 +1,401 @@
+---
+title: 'Validation, Computed Fields, and Custom Logic'
+metaTitle: 'Validation, Computed Fields, Excluding Keys, and Wrapping the Client (Concepts)'
+metaDescription: 'How to add validation, computed fields, and custom logic to applications using the Prisma Client'
+tocDepth: 2
+---
+
+## Extending the Prisma Client
+
+A common request we get is that while Prisma is great for working with the database, your data needs are far more complex. You need to be able to:
+
+- **Validate** that the data that my user sent me is complete and valid
+- Derive **computed fields** based on existing data
+- **Exclude sensitive** **fields** that shouldn't be accessible to your users
+- **Wrap the Prisma Client** to organize and group complex business logic.
+
+Today you'll learn how to implement these patterns in your own application in a type-safe way.
+
+We'll use the following schema to model a simplified SaaS app like Slack or Notion where each Team has many Users. We'll focus on an initial signup flow to illustrate these patterns.
+
+```jsx
+generator client {
+  provider = "prisma-client-js"
+}
+
+datasource db {
+  provider = "postgresql"
+  url      = env("DATABASE_URL")
+}
+
+model User {
+  id        Int    @id @default(dbgenerated())
+  email     String @unique
+  password  String
+  firstName String
+  lastName  String
+  Team      Team?  @relation(fields: [teamId], references: [id])
+  teamId    Int?
+}
+
+model Team {
+  id    Int    @id @default(dbgenerated())
+  name  String @unique
+  users User[]
+}
+```
+
+## Creating a Helper Function
+
+To kick things off, let's discuss how to abstract Prisma Client queries into functions. This will be used in the examples that follow.
+
+Let's build a signup function that requires user and team information to create rows in your database. We'll build this function first in Javascript then in Typescript.
+
+### Helper Function in Javascript
+
+In Javascript, that would look something like following:
```

Comment:
We could probably use a code switcher here

Reply:
Top level language switcher like in "Getting Started" or just tabs?

Reply:
Tabs for index.js / index.ts would be the best I guess, and could be done in the rest of the doc, also removing the lines that say `In Javascript, that would look something like following:`

Reply:
Good call. Cleans up a lot of the article 🙌